src/IO/readers.jl

```diff
@@ -0,0 +1,356 @@
+abstract type AbstractReader end
+
+using HDF5
+using StaticArrays
+using ..Domains: IntervalDomain, SphereDomain
+using ..Meshes:
+    Meshes,
+    NormalizedBilinearMap,
+    IntervalMesh,
+    RectilinearMesh,
+    EquiangularCubedSphere
+using ..Topologies: Topologies, IntervalTopology, Topology2D
+using ..Spaces:
+    Spaces,
+    Spaces.Quadratures,
+    Spaces.Quadratures.GLL,
+    Spaces.CellCenter,
+    Spaces.CellFace,
+    SpectralElementSpace1D,
+    SpectralElementSpace2D,
+    CenterExtrudedFiniteDifferenceSpace,
+    FaceExtrudedFiniteDifferenceSpace,
+    ExtrudedFiniteDifferenceSpace
+using ..Fields: Field, FieldVector
+import ..Geometry:
+    Geometry,
+    XPoint,
+    YPoint,
+    ZPoint,
+    Covariant1Vector,
+    Covariant12Vector,
+    Covariant3Vector
+using ..DataLayouts
+
+"""
+    HDF5Reader
+
+A struct containing the `filename` and `cache` for reading data from the HDF5 file.
+The `cache` enables reading of relevant data without duplication.
+"""
+struct HDF5Reader
+    file::HDF5.File
+    cache::Dict{Any, Any}
+end
+
+function HDF5Reader(filename::AbstractString)
+    file = h5open(filename, "r")
+    cache = Dict()
+    return HDF5Reader(file, cache)
+end
+
+"""
+    Base.read!(reader, path)
+
+Checks the cache to verify if the object, specified by the `path`, is in the cache.
+If it has not been read, this function reads it to the cache and returns the object.
+"""
+function Base.read!(reader::HDF5Reader, path::AbstractString)
+    (; file, cache) = reader
+    name = split(path, "/")[end]
+    if !haskey(cache, name)
+        @assert haskey(file, path)
+        read_new!(reader, name, _type(reader, path))
+    end
+    return cache[name]
+end
+
+"""
+    _type(reader, path)
+
+Extracts the type of the object, specified by the `path`.
+"""
+_type(reader::HDF5Reader, path::AbstractString) =
+    eval(Meta.parse(attrs(reader.file[path])["type"]))
+
+"""
+    matrix_to_cartesianindices(elemorder_matrix)
+
+Converts the `elemorder_matrix` to cartesian indices.
+"""
+function matrix_to_cartesianindices(elemorder_matrix)
+    m, ndims = size(elemorder_matrix)
+    dims = [maximum(elemorder_matrix[:, dim]) for dim in 1:ndims]
+    elemorder = Vector{CartesianIndex{ndims}}(undef, m)
+    for i in 1:m
+        elemorder[i] = CartesianIndex(elemorder_matrix[i, :]...)
+    end
+    return reshape(elemorder, dims...)
+end
+
+"""
+    read_new!(reader, name, ::Type{IntervalDomain})
+
+Reads an object named 'name' of type 'IntervalDomain'.
+"""
+function read_new!(
+    reader::HDF5Reader,
+    name::AbstractString,
+    ::Type{IntervalDomain},
+)
+    g = reader.file["domains/" * name]
+    coord_type = replace(attrs(g)["coord_type"], "ClimaCore." => "")
+    boundary_names =
+        haskey(attributes(g), "boundary_names") ?
+        eval(Meta.parse(attrs(g)["boundary_names"])) : nothing
+    periodic = boundary_names === nothing ? true : false
+    CoordType = eval(Meta.parse(coord_type))
```

Comment:
There are only handful of coord types: can we just make this an if-else branch instead?